### DISCUSSION TITLE
timerCallback() does not seem to be called automatically

### DISCUSSION THREAD

**Discussion by seogyeongjeong** [ANLAB-KAIST/KENSv3#78] [Project Q&A 2. Transport Layer (TCP)] [answered]
I've read about the page of [E::TimerModule](http://anlab-kaist.github.io/KENSv3/doc/class_e_1_1_timer_module.html) a lot of times and still have trouble with handling timer.
I started Timer by calling 
`this->addTimer(NULL, 1000);`

And I implemented print statement inside timerCallback() function like
```
void TCPAssignment::timerCallback(std::any payload) {
  printf("\ntimerCallback() Called!\n");
}
```
right below addTimer() function.

However, the printf statement is never called.

What could be the problem? What document can I read about this?

**Comment by Pusnow** [comment, marked as answer]
It works for me.

Source code:

```c++
void TCPAssignment::initialize() {
  this->addTimer(NULL, 1000);
}

...

void TCPAssignment::timerCallback(std::any payload) {
  printf("\ntimerCallback() Called!\n");
}
```

Output:
```
==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from TestEnv_Reliable
[ RUN      ] TestEnv_Reliable.TestOpen
[RANDOM_SEED : 1614233283 RUN_SOLUTION : 0 UNRELIABLE : 0]

timerCallback() Called!
/home/pusnow/KENSv3/app/kens/testopen.cpp:56: Failure
Expected equality of these values:
  success
    Which is: 0
  ((size_t)test_size) * ((size_t)test_repeat)
    Which is: 131072
/home/pusnow/KENSv3/app/kens/testopen.cpp:57: Failure
Expected equality of these values:
  failed
    Which is: 131072
  0
/home/pusnow/KENSv3/app/kens/testopen.cpp:58: Failure
Expected equality of these values:
  duplicated
    Which is: 67043328
  0
[  FAILED  ] TestEnv_Reliable.TestOpen (158 ms)
```

If you call `addTimer()` inside `packetArrived()`, then you should check `packetArrived()` is called.
`test-kens-open` will not call  `packetArrived()`.

**Reply by seogyeongjeong** [reply]
Hi, I am still struggling with the same issue.
When I use start Timer just before while(1) loop and sees if the timerCallback() is called(by implementing printf on timerCallback() ),
the timerCallback() never seems to be called. 
There is no corresponding printing message shown.

Also, when I started many timers on one 'case', all the corresponding timerCallback()s are called sequentially right after SystemCallback() function is returned.
I can't find a way to wait for the timer within the same block of 'case'.
Can I get some help?

**Reply by Pusnow** [reply]
Oh, since the `while(1)` loop, your implementation never returns to the KENS's scheduler, and the timer is never fired.

You should not use `while(1)` and make sure all callbacks (systemCall, packetArrived, timer, etc.) return.

> Also, when I started many timers on one 'case', all the corresponding timerCallback()s are called sequentially right after SystemCallback() function is returned.

The `timerCallback` behavior is normal.


> I can't find a way to wait for the timer within the same block of 'case'.

You can't. If you wonder how to implement blocking system call, this might help.

ex) Example `sleep_ns()` system call implementation. (Disclaimer: never tested. Might not compile)


```c++
void TCPAssignment::sleep_ns(UUID syscallUUID, int pid, int nanoseconds){
    addTimer(syscallUUID, nanoseconds);
    // Do not call
    // this->returnSystemCall(syscallUUID, 0);
    // to block this system call.
}

void TCPAssignment::timerCallback(std::any payload) {
    UUID syscallUUID = std::any_cast<UUID>(payload);
    this->returnSystemCall(syscallUUID, 0); // unblock system call

}

```
